### ISSUE TITLE
ThreadController static nature

### ISSUE THREAD

**Issue by LukasPaczos** [mapbox/mapbox-navigation-android#3156]
The [`ThreadController` class](https://github.com/mapbox/mapbox-navigation-android/blob/master/libnavigation-util/src/main/java/com/mapbox/navigation/utils/internal/ThreadController.kt) currently operates in a static fashion, there's only one controller for the whole app process. In order to prevent leaks, the `MapboxNavigation` starts the jobs when created and shuts down the controller in the `MapboxNavigation#onDestroy` which joins/kills all of the threads and coroutines.

This introduces an issue whenever the `MapboxNavigation` class needs to be recreated. It heavily depends on the order of the calls and strictly requires the old `MapboxNavigation` instance's `#onDestroy` to be called before the new instance's constructor is called. Otherwise, `MapboxNavigation` is not responsive because all the coroutines are killed.

While we do not expect 2 `MapboxNavigation` instances to be running in parallel, we also should remove the method invocation order that's currently required.

TODO:
- [ ] Update documentation to highlight that only one instance of `MapboxNavigation` is permitted at a time and explaining that the first one needs to be fully cleared before we can create a new one.
- [ ] `[optional]` Introduce a safeguard that keeps a count of `MapboxNavigation` instances count and throws or silently logs if the count is more than one.
- [ ] Remove the static `ThreadController` and replace it with an interface. The implementation instance can be managed internally in the `MapboxNavigation` and injected where necessary.

Do the next actions look good @mapbox/navigation-android?

**Comment by evabishchevich** [comment]
The same is applicable for MapboxTripNotification. https://github.com/mapbox/mapbox-navigation-android/blob/master/libtrip-notification/src/main/java/com/mapbox/navigation/trip/notification/MapboxTripNotification.kt#L75

**Comment by kmadsen** [comment]
> Update documentation to highlight that only one instance of MapboxNavigation is permitted at a time and explaining that the first one needs to be fully cleared before we can create a new one.

During a sync, @cafesilencio and I were discussing similar MapboxNavigation issues; specifically for [handling activity configuration changes](https://github.com/mapbox/navigation-sdks/issues/402).

An idea discussed, was to manage the state and instances with a provider. This would also allow us to retrieve MapboxNavigation instances that have already been created.

For example: Similar to how Android ViewModelProvider works. We could create the api definition with a MapboxProvider

```
val mapboxNavigation = MapboxProvider()
    .navigation(navOptions)

// later down the road it could provide the trip notification 
val tripNotification = MapboxProvider()
    .tripNotification(tripOptions)
```

 - Hide the singleton in MapboxProvider
 - When new options are set, the old navigation is destroyed and the new navigator is created
 - The provider may consider async operations to handle things like map loading

This OP solution does not help the configuration change problem. It's hard to know, from an Activity's perspective, that a MapboxNavigation has already been created. Documentation would still require the developer to build their own version a MapboxNavigation instance manager (named "MapboxNavigationProvider" in this example)

^ This makes me less excited about the current approach. Have alternatives been considered? Agreeing with the ThreadController and [MapboxTripNotification](https://github.com/mapbox/mapbox-navigation-android/issues/3156#issuecomment-642551337) problems. A solution would also help [MapboxNativeNavigatorImpl](https://github.com/mapbox/mapbox-navigation-android/blob/6f5359e789d4aa24ac01e937a98663198bedc8d1/libnavigator/src/main/java/com/mapbox/navigation/navigator/internal/MapboxNativeNavigatorImpl.kt#L42)

**Comment by korshaknn** [comment]
we call ` ThreadController.init()` in `MapboxNavigation init block`
so, maybe just cancel all jobs children here?
```
fun init() {
        cancelAllNonUICoroutines()
        cancelAllUICoroutines()

        ioRootJob = SupervisorJob()
        mainRootJob = SupervisorJob()
    }
```
or we are going to wrap it as mentioned above?
@LukasPaczos , @Guardiola31337

**Comment by korshaknn** [comment]
I think it's a good idea to wrap MabpoxNavigation with Provider.
We need to implement a clear way to
- initialize MabpoxNavigation with options
- get already initialize MabpoxNavigation
- reinitialize MabpoxNavigation (Provider will handle clean up logic)

**Comment by korshaknn** [comment]
closed with #3212
